### PR TITLE
[inductor] Reindex layout consumers before index inversion

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #190596
* #190595
* #190594
* #190593
* #190404
* #190403
* __->__ #190402

Index inversion only analyzes a single logical loop variable. Equal-numel layout consumers can retain a higher-rank iteration domain, causing compatible producer-to-clone paths such as attention transposes, channel shuffles, and reduced scale swizzles to miss fusion before inversion is attempted.

Flatten the pointwise consumer to the producer write domain and retry the existing inversion analysis. Check the symbolically flattened read before rebuilding the LoopBody, then canonicalize the one-variable expression so the preflight and rebuilt consumer share a scheduler-local cached inverse. This preserves the 16 newly enabled ShuffleNet fusions while avoiding repeated symbolic inversion work.

The cache uses the existing per-instance cache helper and has the lifetime of one Scheduler. Caching a failed proof can only miss this optional fusion if ShapeEnv facts strengthen later; it cannot affect correctness. Keep rollback at the existing can_fuse mutation-tracker boundary and reject zero-numel flattened domains before decomposition.

Test Plan:

```
python test/inductor/test_loop_ordering.py TestIndexInversion
CUDA_VISIBLE_DEVICES=0 python test/inductor/test_loop_ordering.py -k reindex
CUDA_VISIBLE_DEVICES=0 python test/inductor/test_nested_reduction.py -k rmsnorm_mxfp8_scale_swizzle
CUDA_VISIBLE_DEVICES=0 python test/inductor/test_custom_lowering.py TestCustomLowering.test_reused_inline_asm_realized
CUDA_VISIBLE_DEVICES=0 python agent_space/bench_shufflenet_reindex.py
https_proxy=http://fwdproxy:8080 lintrunner -a
```

Authored with an AI assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_nested_reduction.py [hunk 1]
```diff
@@ -97,6 +97,24 @@ def _layernorm(x_flat):
...
+
+def _rmsnorm_block_scale_swizzle(x, weight, G):
+    import torch.nn.functional as F
+
+    B, D = x.shape
+    x = F.rms_norm(x, (D,), weight)
+    x_groups = x.view(B, D // G, G)
+    amax = x_groups.abs().amax(dim=-1)
+    scale = (amax / 448.0).clamp(min=1e-12)
+    payload = (x_groups / scale.unsqueeze(-1)).to(torch.float16)
```
Comment (drisspg) [inline test/inductor/test_nested_reduction.py:114]: shoudl this be float8?

**Review by drisspg — APPROVED** [review]
Codex thinks that this identity invert map will fail  

```
identity_expr = (
    128 * FloorDiv(p, 128)
    + ModularIndexing(p, 1, 128)
)
self.assertEqual(
    generate_inverse_formula(identity_expr, p, 16384),
    p,
)

```

File: test/inductor/test_nested_reduction.py (hunk 1 above)
Comment (eellison) [reply test/inductor/test_nested_reduction.py:114]: the fusion is the same, its just more finicky with ulp accuracy etc. so switched to fp16.

**Comment by eellison** [comment]
> identity_expr = (
    128 * FloorDiv(p, 128)
    + ModularIndexing(p, 1, 128)
)

This would just get canonicalizes to `p` anyway. and would have already fused.

**Comment by zou3519** [comment]
This leads to a 18% compile time regression, according to claude using one of the oncall skills:
```
Confirmed Regression

- Benchmark: shufflenet_v2_x1_0, torchbench inference
- Transition: 2026-07-23 07:05:52 → 07:22:49 UTC
- Compile time: 15.23s → 18.08s, +18.7%
- Signal: 40.4× historical CV; 14.5× maximum day-over-day change
- Persistence: 63 subsequent commits remain elevated
- Scope: Isolated. Only 1 of 51 nonzero torchbench inference benchmarks exceeded 5%; suite median was -0.29%.
- Culprit: 67d5640b199acd9dad8affa1b75a8b38462bf566
  (https://github.com/pytorch/pytorch/commit/67d5640b199acd9dad8affa1b75a8b38462bf566), [inductor] Reindex layout
  consumers before index inversion

The commit adds speculative flattening and index-inversion analysis during fusion scoring, including SymPy expansion
and simplification. ShuffleNet repeatedly performs reshape/transpose channel-shuffle operations matching these
multidimensional equal-numel consumers, causing substantially more scheduler analysis. Training increased only 3.6%,
below threshold.
```
The dashboard shows a regression ([link](https://hud.pytorch.org/benchmark/v3/dashboard/compiler_inductor?renderGroupId=main&time.start=2026-07-17T00%3A00%3A00.000Z&time.end=2026-07-24T23%3A59%3A59.999Z&filters.repo=pytorch%2Fpytorch&filters.benchmarkName=compiler&filters.mode=inference&filters.dtype=bfloat16&filters.deviceName=cuda+%28h100%29&filters.device=cuda&filters.arch=h100&filters.suite=torchbench&filters.compiler=cudagraphs&lcommit.commit=abf3cf121f164f1b15b8001fc7ddb4f49f5d1ee0&lcommit.workflow_id=29545944855&lcommit.date=2026-07-17T02%3A00%3A00Z&lcommit.branch=main&rcommit.commit=d1e2802e366c287c4773a50f4f0e8c35e8647bbb&rcommit.workflow_id=30057306858&rcommit.date=2026-07-24T02%3A00%3A00Z&rcommit.branch=main&lbranch=main&rbranch=main&maxSampling=110)); the oncall tool has per-commit data, so it probably has found the right commit, but idk where the data actually is.

Because Elias is OOO I'm going to revert this to be safe and see if the numbers go back
